Firestore updates

pulyaevskiy opened the pull request:
Adds:

- `DocumentReference.delete()`
- `Query.where`
- `Query.orderBy`

Tested in iOS and Android (si|e)mulators.

For the `Query` methods I had to do some string manipulation to fit everything in existing `parameters` Map, but I'd be happy to refactor if needed.

Fixes flutter/flutter#12802 flutter/flutter#12509 flutter/flutter#12396.

pulyaevskiy commented:
Also I noticed there are some unhandled exceptions in Android implementation, I'd like to send those to Dart as Stream errors but I'm not sure what's the API for this, it would be really helpful if someone can point me in the right direction.

kroikie commented:
@pulyaevskiy Thanks so much for the contribution, I would vote for the refactor, using a more structured version of `where`. It always has three parts kind (field or path), operator and value, so the `where` parameters field could contain an object defined as such.

pulyaevskiy commented:
@kroikie 

Thank you for the feedback! If I understand correctly it would be like:

```dart
query.where(new Field('singer', '==', 'Norah Jones'));
```

I can also think of adding some sugar, e.g.:

```dart
query.where(new Field.equalTo('singer', 'Norah Jones'));
query.where(new Field.greaterThan('createdAt', someDate));
// etc.
```

I guess for me it looks a bit verbose but works.
I'm also not sure I really understand difference between regular *Field methods and *FieldPath methods. It looks like you can refer to nested fields using dot-notation in regular *Field methods on the native side (e.g. `[query queryWhereField:@"some.nested.key" isEqualTo:@"value"]`).

Some other options I had in mind:

1. Do something similar to iOS:

```dart
query.where('singer', isEqualTo: 'Norah Jones');
```

I really like how it reads as a sentence. You could also combine range queries in one call because all operators would be in the named arguments. Method singature would be:

```dart
Query where(String field, {
  dynamic isEqualTo, 
  dynamic lessThan, 
  dynamic lessThanOrEqualTo,
  dynamic greaterThan,
  dynamic greaterThanOrEqualTo,
});
```

If distinction between Field and FieldPath is required we can make two methods `whereField` and `whereFieldPath`.

2. Employ `@proxy` and generics and do something more ORM-like (as in Scala Slick library for instance):

```dart
firestore.collection<User>('/users')
  .where((User user) => user.createdAt >= someDateTime)
  .orderBy((User user) => user.createdAt, descending: true);
```

I'm not sure if Flutter is OK with `noSuchMethod()` and `Invocation`. It would also be a lot more involved and opinionated implementation. Also `@proxy` seems to be deprecated now :(.
I played around this concept a year ago [here](https://github.com/corsac-dart/dal/blob/master/lib/src/filter.dart). One nice thing about this is that it provides interface similar to `Stream` and `List`.

Overall, after thinking about this more, I feel like my favorite at this point is the iOS like approach with named arguments, but let me know your thoughts.

mit-mit commented:
For inspiration, the various APIs are here (click each of the language options to see them): https://firebase.google.com/docs/firestore/query-data/queries#compound_queries

I personally think the Swift one looks great, and from a quick chat with the Dart language folks here they also seemed to like that.

pulyaevskiy commented:
Yes, Swift version is what I was referring to as iOS. My favorite so far as well.

kroikie commented:
+1 for using a Swift-like signature, @pulyaevskiy I think you did indeed understand my suggestion. Also I was hoping to avoid using string checks when determining which version of the query methods to use on the native side. Instead of using a string check like `startsWith` I was hoping we could have a more concrete way of examining the parameters like having an object with known fields.

pulyaevskiy commented:
@kroikie 

> Also I was hoping to avoid using string checks

I agree. What if in the `parameters` map we'd have `where` key for all where conditions and `orderBy` for ordering. The value in the `where` key would be a list of conditions and each condition is a tuple3 of [fieldName, operator, value]? This is simplest solution I can think of to avoid string manipulations.
So the parameters map would look something like this:

```dart
Map parameters {
  'where': [
    ['createdAt', '>=', '<start_date>'],
    ['createdAt', '<', '<end_date>'],
  ],
  'orderBy': ['createdAt', true], // tuple2 where second item is direction
};
```

kroikie commented:
@pulyaevskiy Yes something like that would make for safer translations of queries across the method channel.

pulyaevskiy commented:
@kroikie sounds good! Will have some updates tonight.

pulyaevskiy commented:
@kroikie @mit-mit 

Pushed updates with new `where` signature and iOS refactor to handle new parameters.
Will update Android as well soon.

Let me know what you think about new API and parameters format!

pulyaevskiy commented:
Android part has been updated as well.

lschmierer commented:
Since you are currently working on the Firestore plugin anyway, maybe you can add support for `DocumentReference.get()` and `Query.get()` as part of this PR?

That would really help me. Implementing the Android part would be trivial for me, but I have never done any Objective-C programming before.

Skylled commented:
@lschmierer You should implement it for Android and file a new PR. Maybe someone will do the ObjC for you, or you'll come back to it and find the ObjC easier than you think.

pulyaevskiy commented:
Hoping someone is still watching this. I just pushed some updates to resolve conflicts and use the new `isMethodCall` matcher.

I've been trying to nail down this weird behavior with deleting items and it seems to be that this is an issue with underlying platform SDKs... but I'm not sure.

In short: deleting a document (while subscribed to a Query matching this document) results in following sequence of snapshot events with DocumentChanges:

1. DocumentChangeType.removed with `hasPendingWrites: false` and `isFromCache: false` - OK
2. DocumentChangeType.added with `hasPendingWrites: false` and `isFromCache: true` - ???
3. DocumentChangeType.removed with `hasPendingWrites: false` and `isFromCache: false` - ???

Tested with `AnimatedList` and `Dismissible` items: deleted item appears in the list for a second and disappears again.

Wondering if anyone knows what am I doing wrong here?

I will be pushing some more changes this week in regards to SnapshotMetadata and adding some more tests.

Skylled commented:
Is it maybe a race between the widget refreshing and the cache updating with the removal?

pulyaevskiy commented:
I suspected so, but according to docs those metadata updates should not be fired unless explicitly requested.

collinjackson commented:
Thanks so much for your contribution, it looks great. I've published the work you've done so far, because I'm working on a project that depends on it. I've filed #13028 to track your remaining work. Please continue in a new PR.

jonasbark commented:
I'm late in the game but I noticed that the query does not accept an "isEqualTo: null" query. Having a where clause checking for null equality is perfectly valid and works in iOS and Android - I've checked that.

Android part: query = query.whereEqualTo(fieldName, null);

pulyaevskiy commented:
@jonasbark does it match documents without specified field? Looks strange, but I can fix it in a follow-up PR. Or feel free to submit a PR yourself.

pulyaevskiy commented:
Thanks @collinjackson for merging this!

jonasbark commented:
@pulyaevskiy field name is specified. It can be used to filter for documents where the field value must be null, like this:
![image](https://user-images.githubusercontent.com/1151304/32851163-872443ba-ca34-11e7-8b22-9fc2f24708ca.png)

Everything else works fine, thank you!

pulyaevskiy commented:
Oh, this may be a challenge with current API, actually. :)

pulyaevskiy commented:
Maybe only:

```dart
query.where('field', isNull: true);
```

Thoughts?

jonasbark commented:
Looks good and clear enough to me. Another possibility would be this:

`query.where('field', isEqualTo: Null)`
isEqualTo != null is true if you send in Null. Of course it's not obvious enough that you could use Null so I prefer your proposal.

collinjackson commented:
I also prefer `isNull: true` over `isEqualTo: Null`

pulyaevskiy commented:
@jonasbark @collinjackson 
See #275

polbatllo commented:
After some testing, I found out that some queries break resulting snapshots.

As an example, the following query would return wrongly

`static Query usersForTeam(BuildContext context) {
    User user = NinjaSelector.user(context);
    // This is working
    return Firestore.instance.collection('users').orderBy('name');

    // This is NOT working correctly: When used inside a StreamBuilder, the returned query snapshots get cleared after a widget rebuild.
    // Firestore.instance.collection('users').where('current_team_id', isEqualTo: user.current_team_id).orderBy('name');
  }`

tracedaily commented on packages/cloud_firestore/lib/src/query.dart:
```diff
@@ -68,4 +78,24 @@ class Query {
+  /// [field].
+  Query orderBy(String field, {bool descending: false}) {
+    final String key = "orderBy:$field";
+    assert(!_parameters.containsKey(key));
```
is this here for a reason? It's preventing multiple orderBy fields. It seems to work when I take it out...

tracedaily replied:
meaning the assert, line 98...
